client/src/main/scala/org/http4s/client/PoolManager.scala

```diff
@@ -85,30 +107,37 @@ private final class PoolManager[F[_], A <: Connection[F]](
         if (!isClosed) {
           @tailrec
           def go(): Unit =
-            idleQueue.dequeueFirst(_.requestKey == key) match {
+            getConnectionFromQueue(key) match {
               case Some(conn) if !conn.isClosed =>
                 logger.debug(s"Recycling connection: $stats")
                 callback(Right(NextConnection(conn, fresh = false)))
 
               case Some(closedConn) =>
                 logger.debug(s"Evicting closed connection: $stats")
-                allocated -= 1
+                decrConnection(key)
                 go()
 
-              case None if allocated < maxTotal =>
+              case None
+                  if curTotal < maxTotal && allocated.getOrElse(key, 0) < maxConnectionsPerRequestKey(
+                    key) =>
                 logger.debug(s"Active connection not found. Creating new one. $stats")
                 createConnection(key, callback)
 
-              case None if idleQueue.nonEmpty =>
+              case None if idleQueue.get(key).exists(_.nonEmpty) =>
                 logger.debug(
                   s"No connections available for the desired key. Evicting oldest and creating a new connection: $stats")
-                allocated -= 1
-                idleQueue.dequeue().shutdown()
+                decrConnection(key)
+                getConnectionFromQueue(key).get.shutdown()
                 createConnection(key, callback)
 
               case None => // we're full up. Add to waiting queue.
                 logger.debug(s"No connections available.  Waiting on new connection: $stats")
-                waitQueue.enqueue(Waiting(key, callback))
+                if (waitQueue.length <= maxWaitQueueLimit) {
+                  waitQueue.enqueue(Waiting(key, callback))
+                } else {
+                  logger.error(s"Max wait length reached, not scheduling.")
+                  callback(Left(new Exception("Wait queue is full")))
```

Comment:
Should add the `maxWaitQueueLimit` to the message to aid in debugging.

Reply:
Ok